lib.c

```diff
@@ -323,20 +329,35 @@ int osc_init_sdk(struct osc_env *e, const char *profile, unsigned int flag)
 		osc_str_append_string(&e->endpoint, endpoint);
 	}
 
-	if (!e->ak || !e->sk) {
-		fprintf(stderr, "access key and secret key needed\n");
+	if (auth && (!strcmp(auth, "password") || !strcmp(auth, "basic")))
+		e->auth_method = OSC_PASSWORD_METHOD;
+	else if (auth && strcmp(auth, "accesskey")) {
+		fprintf(stderr, "'%s' invalude auth methode\n", auth);
 		return -1;
 	}
 
-	if (strlen(e->ak) != AK_SIZE || strlen(e->sk) != SK_SIZE) {
-		fprintf(stderr, "Wrong access key or secret key size\n");
-		return -1;
+	if (e->auth_method == OSC_AKSK_METHODE) {
+		if (!e->ak || !e->sk) {
+			fprintf(stderr, "access key and secret key needed\n");
+			return -1;
+		}
+
+		if (strlen(e->ak) != AK_SIZE || strlen(e->sk) != SK_SIZE) {
+			fprintf(stderr, "Wrong access key or secret key size\n");
+			return -1;
+		}
+	} else if (e->auth_method == OSC_PASSWORD_METHOD) {
+		e->ak = getenv("OSC_LOGIN");
+		e->sk = getenv("OSC_PASSWORD");
```

Comment:
could you use dedicated struct element like `e->login` and `e->password` to avoid confusion and future errors?

Reply:
I could but that would duplicate the code too.
i could use a
```C
union {
char *ak;
char *login;
}
```
but anonymous union are not C89 and header, try to be C89.